stored_requests/fetcher.go

```diff
@@ -19,52 +19,125 @@ type Fetcher interface {
 // Implementations must be safe for concurrent access by multiple goroutines.
-// To add a Cache layer in front of the Fetcher, see WithCache()
+// To add a Cache layer in front of a Fetcher, see WithCache()
 type Cache interface {
-	// GetRequests works much like Fetcher.FetchRequests, with a few exceptions:
+	// Get works much like Fetcher.FetchRequests, with a few exceptions:
 	//
 	// 1. Any errors should be logged by the implementation, rather than returned.
 	// 2. The returned map _may_ be written to.
 	// 3. The returned map must _not_ contain keys unless they were present in the argument ID list.
 	// 4. Callers _should not_ assume that the returned map contains a key for every id.
 	//    The returned map will miss entries for keys which don't exist in the cache.
-	GetRequests(ctx context.Context, ids []string) map[string]json.RawMessage
+	Get(ctx context.Context, ids []string) map[string]json.RawMessage
 
-	// SaveRequests stores some data in the cache. The map is from ID to the cached value.
-	//
-	// This is a best-effort method. If the cache call fails, implementations should log the error.
-	SaveRequests(ctx context.Context, values map[string]json.RawMessage)
+	// Invalidate will ensure that all values associated with the given IDs
+	// are no longer returned by the cache until new values are saved via Update
+	Invalidate(ctx context.Context, ids []string)
+
+	// Update will update the given values in the cache
+	Update(ctx context.Context, data map[string]json.RawMessage)
+}
+
+type composedCache struct {
+	caches []Cache
+}
+
+// Compose will compose caches into a single cache.
+// The returned cache will attempt to Get from the caches in the order in which. they are provided,
+// stopping as soon as a value is found (or when all caches have been exhausted).
+// Invalidations and updates are propagated to all underlying caches.
+func Compose(caches []Cache) Cache {
+	return &composedCache{
+		caches: caches,
+	}
+}
+
+func (c *composedCache) Get(ctx context.Context, ids []string) (data map[string]json.RawMessage) {
+	data = make(map[string]json.RawMessage)
+	remainingIds := ids
+
+	for _, cache := range c.caches {
+		if cachedData := cache.Get(ctx, remainingIds); len(cachedData) > 0 {
+			// iterate over remainingIds from end, droppings ids as they are filled
+			for i := len(remainingIds) - 1; i >= 0; i-- {
+				if config, ok := cachedData[remainingIds[i]]; ok {
+					data[remainingIds[i]] = config
+					remainingIds = append(remainingIds[:i], remainingIds[i+1:]...)
+				}
+			}
+
+			// return if all ids filled
+			if len(remainingIds) == 0 {
+				return
+			}
+		}
+	}
+	return
+}
+
+func (c *composedCache) Invalidate(ctx context.Context, ids []string) {
+	for _, cache := range c.caches {
+		cache.Invalidate(ctx, ids)
+	}
+}
+
+func (c *composedCache) Update(ctx context.Context, data map[string]json.RawMessage) {
+	for _, cache := range c.caches {
```

Comment:
This will store a copy of the new update into every cache we have, meaning we have N copies in memory where we will only use one. I would suggest checking each cache to see if the ID exists, and only updating the cache if it is found. If the id was not in any of the caches to begin with, we probably don't need to cache the update.

Further optimizing the update would depend on the use case(s) of having multiple caches. My first thought is we should invalidate later entries once we find a copy in the first cache, but that would break expected behavior if the various caches have different TTLs, or other operating parameters which would have the item leave the first cache prior to leaving a later cache. What use case(s) do you have in mind?

Reply:
I don't think having an `Update()` is bad, I can see a possibility along the lines of the CPU L1, L2, L3 ... cache structure. A small(er) cache in memory, a larger one on disk, with a more aggressive timeout on the in-memory cache. But I think we are far from being at a state to work on such optimizations yet. Good to have for when we do want it, but if implementing this PR was being slowed down significantly because of it, I would leave it for a later PR.

On the comparison to the CPU caches, I seem to recall Intel and AMD having two different strategies. One would store the same data in L1 and L2, so it would still be found in L2 after it expires in L1. The other would only store in L1, and made the copy to L2 part of the L1 eviction process. That both are used in a heavily optimized process (CPU architecture) show there is no general best choice on this. :)

Reply:
Really the only use case I had in mind was putting some kind of cache shared by multiple PBS instances (e.g. memcached) between the in-memory LRU cache and the database. I don't anticipate long chains of caches, I just wanted to make the interface reasonably generic.

Checking which keys in an update are in a cache and then updating only those keys in that cache layer adds some overhead, though it's probably acceptable as it's overhead on `Update`, not on `Get`, which should always be as fast as possible. Perhaps the `Cache` implementations themselves can decide how to handle an `Update` with keys that don't exist (e.g. store them or ignore them).

Reply:
This looks good to me as is. First get it working. Then get it working well :)

We can't really prove which strategy makes the most sense until we add more detailed metrics and measure it anyway.